File: bom-to-excel.py
import json
import pandas as pd
import os
import argparse
from openpyxl import Workbook
from openpyxl.styles import Alignment, PatternFill, Font, Border, Side

def classify_category(group, name):
    """프런트엔드 라이브러리 특성에 따른 카테고리 분류 로직"""
    g = str(group).lower() if group else ""
    n = str(name).lower() if name else ""

    if any(x in n for x in ['react', 'zustand', 'router', 'context']): return 'Framework (React/Core)'
    if any(x in n for x in ['tiptap', 'prosemirror']): return 'Editor (RichText/Tiptap)'
    if any(x in n for x in ['vite', 'eslint', 'prettier', 'postcss']): return 'Development (Build/Lint)'
    if any(x in n for x in ['testing', 'vitest', 'jest', 'jsdom']): return 'Development (Testing)'
    if any(x in n for x in ['axios', 'fetch']): return 'Network (API/Http)'
    if 'types' in n or 'types' in g: return 'Development (Types/TS)'
    if any(x in n for x in ['svar-ui', 'grid', 'table']): return 'UI Component (Grid/Table)'

    return 'Other Libraries'

def run_conversion(input_path, output_path):
    if not os.path.exists(input_path):
        print(f"❌ 오류: {input_path} 파일이 없습니다.")
        return

    try:
        # 1. JSON 데이터 읽기
        with open(input_path, 'r', encoding='utf-8') as f:
            data = json.load(f)

        # 2. 데이터 가공
        rows = []
        for comp in data.get('components', []):
            licenses = comp.get('licenses', [])
            license_str = "Unknown"
            if licenses:
                license_list = []
                for l in licenses:
                    lic = l.get('license', {})
                    name = lic.get('id') or lic.get('name') or l.get('expression')
                    if name: license_list.append(name)
                if license_list: license_str = ", ".join(license_list)

            rows.append({
                'Category': classify_category(comp.get('group'), comp.get('name')),
                'Package Name': comp.get('name'),
                'Installed Version': comp.get('version'),
                'License': license_str,
                'Description': comp.get('description') or comp.get('name')
            })

        # 3. 데이터프레임 생성 및 정렬
        df = pd.DataFrame(rows)
        df = df.sort_values(by=['Category', 'Package Name'])

        # 4. 엑셀 생성 및 스타일링
        with pd.ExcelWriter(output_path, engine='openpyxl') as writer:
            df.to_excel(writer, index=False, sheet_name='SBOM')
            ws = writer.sheets['SBOM']

            # 헤더 고정
            ws.freeze_panes = 'A2'

            # 스타일 정의
            header_fill = PatternFill(start_color="B4C6E7", end_color="B4C6E7", fill_type="solid")
            border = Border(left=Side(style='thin'), right=Side(style='thin'),
                            top=Side(style='thin'), bottom=Side(style='thin'))
            align_center = Alignment(horizontal='center', vertical='center', wrap_text=True)
            align_left = Alignment(horizontal='left', vertical='center', wrap_text=True)

            # 헤더 및 데이터 셀 스타일/병합 적용
            for cell in ws[1]:
                cell.fill = header_fill
                cell.font = Font(bold=True)
                cell.alignment = align_center
                cell.border = border

            current_cat, start_row = "", 2
            max_row = ws.max_row
            for i in range(2, max_row + 1):
                for col in range(1, 6):
                    c = ws.cell(row=i, column=col)
                    c.border = border
                    c.alignment = align_center if col <= 4 else align_left

                cat_val = ws.cell(row=i, column=1).value
                if cat_val != current_cat:
                    if i > start_row + 1:
                        ws.merge_cells(start_row=start_row, start_column=1, end_row=i-1, end_column=1)
                    current_cat, start_row = cat_val, i

            if max_row >= start_row and max_row > 2:
                ws.merge_cells(start_row=start_row, start_column=1, end_row=max_row, end_column=1)

            # 너비 설정
            ws.column_dimensions['A'].width = 25
            ws.column_dimensions['B'].width = 35
            ws.column_dimensions['C'].width = 15
            ws.column_dimensions['D'].width = 20
            ws.column_dimensions['E'].width = 60

        print(f"✅ 성공: {output_path} 생성 완료!")

    except Exception as e:
        print(f"❌ 오류 발생: {e}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Convert CycloneDX bom.json to Styled Excel')
    parser.add_argument('--input', '-i', default='bom.json', help='Input JSON file path (default: bom.json)')
    parser.add_argument('--output', '-o', default='SMEP_AFE_Architecture_BOM.xlsx', help='Output Excel file path')

    args = parser.parse_args()
    run_conversion(args.input, args.output)

File: bom-to-excel.js
import fs from 'fs';
import * as XLSX from 'xlsx';

// 1. 경로 설정
const inputJson = './bom.json';
const outputExcel = './SMEP_AFE_Architecture_BOM.xlsx';

// 2. 카테고리 분류 로직 (Python 스크립트 방식 참고 및 프론트엔드 최적화)
function classifyCategory(group, name) {
    const g = (group || "").toLowerCase();
    const n = (name || "").toLowerCase();

    if (n.includes('react') || n.includes('zustand') || n.includes('router')) return 'Framework (React/Core)';
    if (n.includes('tiptap') || n.includes('prosemirror')) return 'Editor (RichText/Tiptap)';
    if (n.includes('vite') || n.includes('eslint') || n.includes('prettier') || n.includes('postcss')) return 'Development (Build/Lint)';
    if (n.includes('testing') || n.includes('vitest') || n.includes('jest') || n.includes('jsdom')) return 'Development (Testing)';
    if (n.includes('axios') || n.includes('fetch')) return 'Network (API/Http)';
    if (n.includes('types') || g.includes('types')) return 'Development (Types/TS)';
    if (n.includes('svar-ui') || n.includes('grid')) return 'UI Component (Grid/Table)';

    return 'Other Libraries';
}

if (!fs.existsSync(inputJson)) {
    console.error(`❌ 오류: ${inputJson} 파일이 없습니다.`);
} else {
    const bomData = JSON.parse(fs.readFileSync(inputJson, 'utf8'));

    // 데이터 가공 및 정렬
    const excelData = bomData.components.map(comp => {
        const category = classifyCategory(comp.group, comp.name);
        let licenseInfo = 'Unknown';
        if (comp.licenses && comp.licenses.length > 0) {
            licenseInfo = comp.licenses
                .map(l => l.license?.id || l.license?.name || l.expression || 'Unknown')
                .join(', ');
        }

        return {
            'Category': category,
            'Package Name': comp.name,
            'Installed Version': comp.version,
            'License': licenseInfo,
            'Description': comp.description || comp.name
        };
    }).sort((a, b) => a.Category.localeCompare(b.Category) || a['Package Name'].localeCompare(b['Package Name']));

    // 엑셀 워크시트 생성
    const worksheet = XLSX.utils.json_to_sheet(excelData);

    // 컬럼 너비 설정
    worksheet['!cols'] = [
        { wch: 25 }, // Category
        { wch: 35 }, // Package Name
        { wch: 15 }, // Installed Version
        { wch: 20 }, // License
        { wch: 60 }  // Description
    ];

    // [스타일링 대안] XLSX 라이브러리 기본 버전은 인라인 스타일링이 제한적이므로
    // 셀 병합(Merging) 정보를 계산하여 추가합니다.
    const mergeCells = [];
    let startRow = 1; // 0-based, 1은 데이터의 첫 줄
    let currentCat = excelData[0].Category;

    for (let i = 1; i < excelData.length; i++) {
        if (excelData[i].Category !== currentCat) {
            if (i - startRow > 1) {
                mergeCells.push({ s: { r: startRow, c: 0 }, e: { r: i, c: 0 } });
            }
            currentCat = excelData[i].Category;
            startRow = i + 1;
        }
    }
    // 마지막 그룹 병합
    if (excelData.length - startRow > 0) {
        mergeCells.push({ s: { r: startRow, c: 0 }, e: { r: excelData.length, c: 0 } });
    }
    worksheet['!merges'] = mergeCells;

    // 워크북 저장
    const workbook = XLSX.utils.book_new();
    XLSX.utils.book_append_sheet(workbook, worksheet, 'SBOM');
    XLSX.writeFile(workbook, outputExcel);

    console.log(`✅ 성공: ${outputExcel} 생성 완료! (카테고리 분류 및 병합 적용)`);
}